CFE-2780: Fixed some of the warnings from autoreconf

larsewi opened the pull request:
[![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=11815)](https://ci.cfengine.com/job/pr-pipeline/11815/)

larsewi committed:
Enabled subdir-objects in AM_INIT_AUTOMAKE

This way we don't get annoying warnings like:

```
warning: source file '$(srcdir)/../../libutils/statistics.c' is in a subdirectory,
but option 'subdir-objects' is disabled
```

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fixed warning AC_PROG_LEX without either yywrap or noyywrap is obsolete

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Ran autoupdate on configure.ac

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fixed warnings after autoupdate

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Upgrade github CI to Ubuntu 24

Ubuntu 20 does not have autoconf version 2.71 or higher

Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Put AC_PREREQ back to version 2.63

We need configure.ac to be supported on older platforms

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

craigcomstock reviewed (dismissed):
right on. should be fine I just have an ongoing maintenance concern about putting lists of packages needed to build all over the place and losing track of some of the places. Can be a follow up ticket for sure.

craigcomstock commented on .github/workflows/asan_unit_tests.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
         submodules: recursive
     - name: Install dependencies
-      run: sudo apt-get update -y && sudo apt-get install -y libssl-dev libpam0g-dev liblmdb-dev byacc curl
+      run: sudo apt-get update -y && sudo apt-get install -y libssl-dev libpam0g-dev liblmdb-dev byacc curl autoconf-archive
```
https://northerntech.atlassian.net/browse/ENT-12578

craigcomstock commented on configure.ac:
```diff
@@ -28,7 +28,7 @@ dnl # Run ./autogen.sh to build configure script
 dnl ##########################################################################
 
-AC_PREREQ([2.71])
+AC_PREREQ([2.63])
```
so squash these commits so the change never happened right?

larsewi replied:
Yeah, I can do that. My initial thought was `autoupdate` should be one commit, then I fix the things that `autoupdate` does in later ones. This way I show the workflow in the commit history. But I guess it does not matter that much.

craigcomstock replied:
that does beg the question actually... as you wish, either way is fine by me. Maybe add an explanation in both commits about your reasoning.

olehermanse commented:
@cf-bottom test this in Jenkins, please :)

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=11610)](https://ci.cfengine.com//job/pr-pipeline/11610/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/11610/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-11610/

larsewi committed:
Fixed failures in distclean target

Building stuff from libutils in test/unit causes make distclean to
delete the same files twice resulting in failure. This happens as an
effect of enabling subdir-objects which causes recursive builds.

Ticket: CFE-2780
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime commented on tests/unit/file_lib_test.c:
```diff
@@ -27,6 +27,8 @@
 #include <file_lib.h>
 #include <stdbool.h>
 
+#include "../../libutils/file_lib.c"
```
Why do we need this? If there are missing function declarations in the `file_lib.h` header, we should add them here.

larsewi replied:
I don't remember right now. Removed it temporarily to remind me :wink:

larsewi replied:
`safe_open()` does not work unless I include the file here or in the `file_lib_test_SOURCES` and I'm not smart enough to know why :thinking: 

The former causes problems with make target distclean on the bootstrap host when subdir_objects is enabled.

vpodzime replied:
Strange, but not worth spending extra time, AFAICT. :+1:

vpodzime reviewed (commented):
Looks good to me otherwise. As long as it works everywhere. :)

vpodzime commented:
Btw, this is the output from `autogen.sh` on my system with these changes:
```
$ NO_CONFIGURE=1 ./autogen.sh 
./autogen.sh: Running autoreconf ...
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:137: warning: The macro `AC_LANG_C' is obsolete.
configure.ac:137: You should run autoupdate.
./lib/autoconf/c.m4:72: AC_LANG_C is expanded from...
m4/acinclude.m4:3: ACX_PTHREAD is expanded from...
configure.ac:137: the top level
configure.ac:137: warning: The macro `AC_TRY_LINK' is obsolete.
configure.ac:137: You should run autoupdate.
./lib/autoconf/general.m4:2920: AC_TRY_LINK is expanded from...
m4/acinclude.m4:3: ACX_PTHREAD is expanded from...
configure.ac:137: the top level
configure.ac:91: installing './compile'
configure.ac:34: installing './config.guess'
configure.ac:34: installing './config.sub'
configure.ac:55: installing './install-sh'
configure.ac:55: installing './missing'
libcompat/Makefile.am: installing './depcomp'
```

larsewi commented:
> Btw, this is the output from autogen.sh on my system with these changes:

Maybe the `ACX_PTHREAD` macro is expanded even though it is in the case/esac ?

larsewi commented:
@cf-bottom Jenkins with exotics please :)

larsewi commented:
@cf-bottom Jenkins with exotics please :)

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=11878)](https://ci.cfengine.com//job/pr-pipeline/11878/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/11878/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-11878/

vpodzime reviewed (approved):
ACK. Thanks!
